### ISSUE TITLE
Table reference in foreignKeys uses working directory as base URL

### ISSUE THREAD

**Issue by bencomp** [ruby-rdf/rdf-tabular#9]
I have been using `rdf validate ...` with a local CSVW-formatted JSON file to check CSV contents. The specific command is

    rdf validate --input-format tabular --metadata data/output/all_owners.csv-metadata.json

This metadata file's `url` property is `all_owners.csv` and there is no `@base` in the `@context`. The file all_owners.csv is in the same directory as the metadata file. This command works as expected.

However, after adding the `foreignKeys` definition

```json
"foreignKeys": [
   {
    "columnReference": "MS_ID",
    "reference": {
     "resource": "all_manuscripts.csv",
     "columnReference": "MS_ID"
    }
   }
  ]
```

the same command exits with

```
ERROR   Schema has invalid property 'foreignKeys': table referenced by file:data/output/all_manuscripts.csv not found
Traceback (most recent call last):
	23: from /usr/local/bin/rdf:23:in `<main>'
	22: from /usr/local/bin/rdf:23:in `load'
	21: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/bin/rdf:13:in `<top (required)>'
	20: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/cli.rb:498:in `exec'
	19: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/cli.rb:634:in `parse'
	18: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/cli.rb:634:in `new'
	17: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/reader.rb:67:in `initialize'
	16: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/reader.rb:302:in `initialize'
	15: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/reader.rb:302:in `instance_eval'
	14: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/reader.rb:88:in `block in initialize'
	13: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/util/logger.rb:191:in `log_depth'
	12: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/util/logger.rb:257:in `log_depth'
	11: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/reader.rb:147:in `block (2 levels) in initialize'
	10: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/cli.rb:635:in `block in parse'
	 9: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/cli.rb:499:in `block in exec'
	 8: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/mixin/mutable.rb:71:in `<<'
	 7: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/mixin/writable.rb:28:in `<<'
	 6: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/mixin/writable.rb:96:in `insert_reader'
	 5: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/mixin/writable.rb:138:in `insert_statements'
	 4: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/reader.rb:167:in `each_statement'
	 3: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/util/logger.rb:191:in `log_depth'
	 2: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-3.0.2/lib/rdf/util/logger.rb:257:in `log_depth'
	 1: from /usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/reader.rb:170:in `block in each_statement'
/usr/local/lib/ruby/gems/2.5.0/gems/rdf-tabular-2.2.1/lib/rdf/tabular/metadata.rb:563:in `validate!': Metadata error (RDF::ReaderError)
```

The file all_manuscripts.csv does exist in the same directory as all_contents.csv and its metadata file. I'm unsure why it's not found.

When I `cd` into `data/output` and run `rdf validate --input-format tabular --metadata ./all_owners.csv-metadata.json`, I get a similar error:

    ERROR   Schema has invalid property 'foreignKeys': table referenced by file:all_owners.csv./all_manuscripts.csv not found

There is something odd about the way file paths are constructed, I think. Or am I doing something wrong?

**Comment by gkellogg** [comment]
I haven't worked on this in a bit, and I recall that the local paths issue was a pain. I'll let you know when I have a chance to look into it further.

**Comment by bencomp** [comment]
Thanks! I'm glad it's not just me. If I can help with debug logs or testing, I'd be very happy to.

**Comment by bencomp** [comment]
I found a work-around for rdf-tabular not finding the tables: use a single metadata.json file that describes the `tableGroup` and combines the metadata from both files. This is an excerpt from `all.csv-metadata.json`:


    "tables": [
        {
            "dc:title": "All texts in the manuscripts",
            "dc:modified": {
                "@value": "2018-09-06",
                "@type": "xsd:date"
            },
            "url": "all_contents.csv",
            "tableSchema": {
                "aboutUrl": "{_row}",
                "primaryKey": [
                    "MS_ID",
                    "item"
                ],
                "columns": [
                    ...
                ],
                "foreignKeys": [
                    {
                        "columnReference": "MS_ID",
                        "reference": {
                            "resource": "all_manuscripts.csv",
                            "columnReference": "MS_ID"
                        }
                    }
                ]
            }
        },
        {
            "dc:title": "Overview of manuscripts and the use of various languages in them",
            "dc:modified": {
                "@value": "2018-09-06",
                "@type": "xsd:date"
            },
            "url": "all_manuscripts.csv",
            "tableSchema": {
                "aboutUrl": "{_row}",
                "primaryKey": "MS_ID",
                "columns": [
                    ...
                ]
            }
        }
    ]

This does not throw errors about not being able to find tables, when I use the CLI in either of these ways:

- `rdf validate --input-format tabular all.csv-metadata.json`
- `rdf validate --input-format tabular --metadata ./all.csv-metadata.json`

In the second command, I had to add `./` to the path of the metadata file, or the command would fail with:
```
/usr/local/lib/ruby/gems/2.6.0/gems/rdf-3.0.12/lib/rdf/util/file.rb:312:in `initialize': No such file or directory @ rb_sysopen - all.csv-metadata.jsonall_contents.csv (Errno::ENOENT)
```

I'm happy now, except… I now discover that rdf-tabular does not check foreign key constraints like a DBMS would: if the `MS_ID` column in `all_contents.csv` contains `1bla`, which is not in the `MS_ID` column in `all_manuscripts.csv`, the input is still valid. This is another issue, I think.

**Comment by gkellogg** [comment]
Apparently the `validate` command isn't enough, you need to do a command which iterates through all the rows, such as `count` and use the `--validate` option to get foreign key checking to work. I'll address that.

The error message is not very useful, so I'll improve that as well.

**Comment by gkellogg** [comment]
As it ends up, the `validate` command isn't enough on its own. This validates the triples that result from parsing the data. To do parse-time validation, you need to include the `--validate` option. The documentation for the `rdf` command can be updated to clarify this in a forthcoming release.

Using the `--validate` option will give you row-by row messages about bad foreign key references.

I have improved some of the error messages. Note, as it happens, the column metadata needs to include both `name` and `titles` properties, due to a working group decision.

**Comment by gkellogg** [comment]
This should be fixed with the 2.2.2 release.
